src/main/java/io/cdap/plugin/gcp/bigquery/connector/BigQueryConnector.java

```diff
@@ -0,0 +1,370 @@
+      browseDetailBuilder
+        .addEntity(BrowseEntity.builder(name, parentPath + name, ENTITY_TYPE_TABLE).canSample(true).build());
+      count++;
+      if (count == countLimit) {
+        break;
+      }
+    }
+    return browseDetailBuilder.setTotalCount(count).build();
+  }
+
+  private BrowseDetail listDatasets(BigQuery bigQuery, Integer limit) {
+    Page<Dataset> datasetPage = bigQuery.listDatasets(BigQuery.DatasetListOption.all());
+    int countLimit = limit == null ? Integer.MAX_VALUE : limit;
+    int count = 0;
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    for (Dataset dataset : datasetPage.iterateAll()) {
+      String name = dataset.getDatasetId().getDataset();
+      browseDetailBuilder
+        .addEntity(BrowseEntity.builder(name, "/" + name, ENTITY_TYPE_DATASET).canBrowse(true).build());
+      count++;
+      if (count == countLimit) {
+        break;
+      }
+    }
+    return browseDetailBuilder.setTotalCount(count).build();
+  }
+
+  private BigQuery getBigQuery() throws IOException {
+    GoogleCredentials credentials = null;
+    //validate service account
+    if (config.isServiceAccountJson() || config.getServiceAccountFilePath() != null) {
+      credentials =
+        GCPUtils.loadServiceAccountCredentials(config.getServiceAccount(), config.isServiceAccountFilePath());
+    }
+    ;
+    return GCPUtils.getBigQuery(config.getDatasetProject(), credentials);
+  }
+
+  private List<StructuredRecord> getData(BigQuery bigQuery, String dataset, String table, int limit)
+    throws IOException {
+    String query = String.format("SELECT * FROM `%s.%s.%s` LIMIT %d", config.getDatasetProject(), dataset, table,
+      limit);
+    QueryJobConfiguration queryConfig = QueryJobConfiguration.newBuilder(query).build();
+    JobId jobId = JobId.of(UUID.randomUUID().toString());
+    Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+    // Wait for the job to finish
+    try {
+      queryJob = queryJob.waitFor();
+    } catch (InterruptedException e) {
+      throw new IOException("Query job interrupted.", e);
+    }
+
+    // check for errors
+    if (queryJob == null) {
+      throw new IOException("Job no longer exists.");
+    } else if (queryJob.getStatus().getError() != null) {
+      throw new IOException(String.format("Failed to query table : %s", queryJob.getStatus().getError().toString()));
+    }
+
+    // Get the results
+    TableResult result = null;
+    try {
+      result = queryJob.getQueryResults();
+    } catch (InterruptedException e) {
+      throw new IOException("Query results interrupted.", e);
+    }
+    List<StructuredRecord> samples = new ArrayList<>();
+
+    com.google.cloud.bigquery.Schema schema = result.getSchema();
+    Schema cdapSchema = BigQueryUtil.getTableSchema(schema, null);
+
+    FieldList fields = schema.getFields();
+    for (FieldValueList fieldValues : result.iterateAll()) {
```

Comment:
would be good to pull out this logic into a separate class that is only responsible for converting FieldValueList into a StructuredRecord. Would also make it easier to unit test.

Reply:
done :)